src/test/deploy/functions/validate.spec.ts

```diff
@@ -0,0 +1,137 @@
+        validate.functionsDirectoryExists("cwd", "sourceDirName");
+      }).to.throw(FirebaseError);
+    });
+  });
+
+  describe("functionNamesAreValid", () => {
+    it("should allow properly formatted function names", () => {
+      const properNames = { "my-function-1": "some field", "my-function-2": "some field" };
+
+      expect(() => {
+        validate.functionNamesAreValid(properNames);
+      }).to.not.throw();
+    });
+
+    it("should throw error on improperly formatted function names", () => {
+      const properNames = {
+        "my-function-!@#$%": "some field",
+        "my-function-!@#$!@#": "some field",
+      };
+
+      expect(() => {
+        validate.functionNamesAreValid(properNames);
+      }).to.throw(FirebaseError);
+    });
+
+    it("should throw error if some function names are improperly formatted", () => {
+      const properNames = { "my-function$%#": "some field", "my-function-2": "some field" };
+
+      expect(() => {
+        validate.functionNamesAreValid(properNames);
+      }).to.throw(FirebaseError);
+    });
+
+    // I think it should throw error here but it doesn't error on empty or even undefined functionNames.
+    // TODO(b/131331234): fix this test when validation code path is fixed.
+    it.skip("should throw error on empty function names", () => {
+      const properNames = {};
+
+      expect(() => {
+        validate.functionNamesAreValid(properNames);
+      }).to.throw(FirebaseError);
+    });
+  });
+
+  describe("packageJsonIsValid", () => {
+    const sandbox: sinon.SinonSandbox = sinon.createSandbox();
+    let cjsonLoadStub: sinon.SinonStub;
+    let fileExistsStub: sinon.SinonStub;
+
+    beforeEach(() => {
+      fileExistsStub = sandbox.stub(fsutils, "fileExistsSync");
+      cjsonLoadStub = sandbox.stub(cjson, "load");
+    });
+
+    afterEach(() => {
+      sandbox.restore();
+    });
+
+    it("should throw error if package.json file is missing", () => {
+      fileExistsStub.withArgs("sourceDir/package.json").returns(false);
+
+      expect(() => {
+        validate.packageJsonIsValid("sourceDirName", "sourceDir", "projectDir");
+      }).to.throw(FirebaseError, "No npm package found");
+    });
+
+    it("should throw error if functions source file is missing", () => {
+      cjsonLoadStub.returns({ name: "my-project" });
+      fileExistsStub.withArgs("sourceDir/package.json").returns(true);
+      fileExistsStub.withArgs("sourceDir/index.js").returns(false);
+
+      expect(() => {
+        validate.packageJsonIsValid("sourceDirName", "sourceDir", "projectDir");
+      }).to.throw(FirebaseError, "does not exist, can't deploy Firebase Functions");
+    });
+
+    it("should throw error if main is defined and that file is missing", () => {
+      cjsonLoadStub.returns({ name: "my-project", main: "src/main.js" });
+      fileExistsStub.withArgs("sourceDir/package.json").returns(true);
+      fileExistsStub.withArgs("sourceDir/srcmain.js").returns(false);
```

Comment:
should this be `sourceDir/src/main.js`

Reply:
Oh yeah you're totally right. Correcting (although you're right that it won't change the test output, but still good to keep things tidy).